src/Microsoft.AspNet.Mvc.Core/HttpResponseStreamWriter.cs

```diff
@@ -0,0 +1,280 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.Framework.Internal;
+
+namespace Microsoft.AspNet.Mvc
+{
+    /// <summary>
+    /// Writes to the <see cref="Stream"/> using the supplied <see cref="Encoding"/>.
+    /// It does not write the BOM and also does not close the stream.
+    /// </summary>
+    public class HttpResponseStreamWriter : TextWriter
+    {
+        private const int DefaultBufferSize = 1024;
+        private readonly Stream _stream;
+        private Encoder _encoder;
+        private byte[] _byteBuffer;
+        private char[] _charBuffer;
+        private int _charBufferSize;
+        private int _charBufferCount;
+
+        public HttpResponseStreamWriter(Stream stream, Encoding encoding)
+            : this(stream, encoding, DefaultBufferSize)
+        {
+        }
+
+        public HttpResponseStreamWriter([NotNull] Stream stream, [NotNull] Encoding encoding, int bufferSize)
+        {
+            this._stream = stream;
+            Encoding = encoding;
+            _encoder = encoding.GetEncoder();
+            _charBufferSize = bufferSize;
+            _charBuffer = new char[bufferSize];
+            _byteBuffer = new byte[encoding.GetMaxByteCount(bufferSize)];
+        }
+
+        public override Encoding Encoding { get; }
+
+        public override void Write(char value)
+        {
+            if (_charBufferCount == _charBufferSize)
+            {
+                Flush();
+            }
+
+            _charBuffer[_charBufferCount] = value;
+            _charBufferCount++;
+        }
+
+        public override void Write(char[] values)
+        {
+            if (values == null)
+            {
+                return;
+            }
+
+            Write(values, 0, values.Length);
+        }
+
+        public override void Write(char[] buffer, int index, int count)
+        {
+            if (buffer == null)
+            {
+                return;
+            }
+
+            while (count > 0)
+            {
+                if (_charBufferCount == _charBufferSize)
+                {
+                    Flush();
+                }
+
+                int remaining = _charBufferSize - _charBufferCount;
+                if (remaining > count)
+                {
+                    remaining = count;
+                }
+
+                Buffer.BlockCopy(
```

Comment:
This block of code starting from the `if` could be moved into a separate method so you get to avoid repeating it in the async code block.